secrets/azurekeyvault/akv.go

```diff
@@ -0,0 +1,187 @@
+//  - keyVersion: string representing the keyVersion, see https://docs.microsoft.com/en-us/rest/api/keyvault/encrypt/encrypt#uri-parameters
+//  - opts: *KeeperOptions with the desired Algorithm to use for operations. See this link for more info: https://docs.microsoft.com/en-us/rest/api/keyvault/encrypt/encrypt#jsonwebkeyencryptionalgorithm
+func NewKeeper(client *keyvault.BaseClient, keyVaultName, keyName, keyVersion string, opts *KeeperOptions) *secrets.Keeper {
+	return secrets.NewKeeper(&keeper{
+		client:       client,
+		keyVaultName: keyVaultName,
+		keyName:      keyName,
+		keyVersion:   keyVersion,
+		options:      opts,
+	})
+}
+
+// Encrypt encrypts the plaintext into a ciphertext.
+func (k *keeper) Encrypt(ctx context.Context, plaintext []byte) ([]byte, error) {
+	if err := k.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	b64Text := base64.StdEncoding.EncodeToString(plaintext)
+	keyOpsResult, err := k.client.Encrypt(ctx, k.getKeyVaultURI(), k.keyName, k.keyVersion, keyvault.KeyOperationsParameters{
+		Algorithm: keyvault.JSONWebKeyEncryptionAlgorithm(k.options.Algorithm),
+		Value:     &b64Text,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(*keyOpsResult.Result), nil
+}
+
+// Decrypt decrypts the ciphertext into a plaintext.
+func (k *keeper) Decrypt(ctx context.Context, ciphertext []byte) ([]byte, error) {
+	if err := k.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	cipherval := string(ciphertext)
+	keyOpsResult, err := k.client.Decrypt(ctx, k.getKeyVaultURI(), k.keyName, k.keyVersion, keyvault.KeyOperationsParameters{
+		Algorithm: keyvault.JSONWebKeyEncryptionAlgorithm(k.options.Algorithm),
+		Value:     &cipherval,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return base64.StdEncoding.DecodeString(*keyOpsResult.Result)
+}
+
+// ErrorAs implements driver.Keeper.ErrorAs.
+func (k *keeper) ErrorAs(err error, i interface{}) bool {
+	e, ok := err.(autorest.DetailedError)
+	if !ok {
+		return false
+	}
+	p, ok := i.(*autorest.DetailedError)
+	if !ok {
+		return false
+	}
+	*p = e
+	return true
+}
+
+// ErrorCode implements driver.ErrorCode.
+func (k *keeper) ErrorCode(err error) gcerrors.ErrorCode {
+	de, ok := err.(autorest.DetailedError)
+	if !ok {
+		return gcerr.Unknown
+	}
+	ec, ok := errorCodeMap[de.StatusCode.(int)]
+	if !ok {
+		return gcerr.Unknown
+	}
+	return ec
+}
+
+func (k *keeper) getKeyVaultURI() string {
+	return fmt.Sprintf("https://%s.%s/", k.keyVaultName, keyVaultEndpointSuffix)
+}
+
+func (k *keeper) validateOptions() error {
+	if k.options != nil && k.options.Algorithm == "" {
+		return fmt.Errorf("azure secret: please select a valid algorithm from (%s)", getSupportedAlgorithmsForError())
+	}
+
+	return nil
+}
+
+func getSupportedAlgorithmsForError() string {
+	algos := ""
+	for _, a := range keyvault.PossibleJSONWebKeyEncryptionAlgorithmValues() {
+		if algos == "" {
+			algos = fmt.Sprintf("%s", a)
```

Comment:
Use `strings.Builder` instead of `fmt.Sprintf`.

Reply:
seems like overkill for a relatively small list of algos (3); strings.Join concatenates - I'll use strings.Join as its less verbose

Reply:
Didn't realize the list is small, sounds good.